desci-server/src/services/data/externalCidProcessing.ts

```diff
@@ -0,0 +1,213 @@
+    await ensureUniquePathsDraftTree({
+      nodeId: node.id,
+      contextPath,
+      externalUrlFilePaths: cidFileNames,
+    });
+
+    /**
+     * Discover the entire tree if it's a DAG, prepare entries for the draft tree
+     */
+    const externalCidMap = {};
+    let entriesDiscovered = [];
+    // const externalDagsToPin = [];
+    if (externalCids?.length && Object.keys(cidTypesSizes)?.length) {
+      entriesDiscovered = [];
+      for await (const extCid of externalCids) {
+        const { size, isDirectory } = cidTypesSizes[extCid.cid];
+        externalCidMap[extCid.cid] = { size, directory: isDirectory, path: extCid.name };
+        if (isDirectory) {
+          //Get external dag tree, add to external dag pin list
+          let tree: RecursiveLsResult[];
+          try {
+            tree = await pubRecursiveLs(extCid.cid, extCid.name);
+          } catch (e) {
+            logger.info(
+              { extCid },
+              '[UPDATE ADD EXTERNAL CID] External DAG tree resolution failed, the contents within the DAG were unable to be retrieved, rejecting update.',
+            );
+            throw createIpfsUnresolvableError(
+              `Failed resolving external dag tree, the DAG or its contents were unable to be retrieved`,
+            );
+          }
+          const flatTree = recursiveFlattenTree(tree);
+          (flatTree as RecursiveLsResult[]).forEach((file: RecursiveLsResult) => {
+            cidTypesSizes[file.cid] = { size: file.size, isDirectory: file.type === FileType.DIR };
+            entriesDiscovered.push({
+              path: contextPath + '/' + file.path,
+              cid: file.cid,
+              size: file.size,
+              isDirectory: file.type === FileType.DIR,
+            });
+            externalCidMap[file.cid] = {
+              size: file.size,
+              directory: file.type === FileType.DIR,
+              path: contextPath + '/' + file.path,
+            };
+          });
+        }
+        // Append root
+        entriesDiscovered.push({
+          path: contextPath + '/' + extCid.name,
+          cid: extCid.cid,
+          size,
+          isDirectory,
+        });
+      }
+    }
+
+    const newDraftTreeEntries: Prisma.DraftNodeTreeCreateManyInput[] = entriesDiscovered.map((entry) => {
+      return {
+        nodeId: node.id,
+        userId: user.id,
+        path: entry.path,
+        cid: entry.cid,
+        size: entry.size,
+        directory: entry.isDirectory,
+        external: true,
+      };
+    });
+
+    const draftTreeEntriesAdded = await prisma.draftNodeTree.createMany({
+      data: newDraftTreeEntries,
+      skipDuplicates: true,
+    });
+    logger.info(`[EXTERNAL CID] ${draftTreeEntriesAdded.count} entries added to draft tree`);
+
+    //repull of node required, previous manifestUrl may already be stale
+    const ltsNode = await prisma.node.findFirst({
+      where: {
+        ownerId: user.id,
+        uuid: node.uuid,
+      },
+    });
```

Comment:
Not sure, but this smells a bit like speeding to win a race condition. Not sure about the context, but do you think the issue is avoided?